### ISSUE TITLE
`sample` test case is broken with 1.75

### ISSUE THREAD

**Issue by osiewicz** [pacak/cargo-show-asm#230]
In PR #229 we started seeing weird CI failures for https://github.com/pacak/cargo-show-asm/blob/89551909865dc3623cc25507fd48e875173d9822/.github/workflows/check-and-lint.yaml#L40 . I suspect that an upgrade from 1.74 to 1.75 is what caused this, as it looks like the CI grabs the newest Rust version automatically. 
Attached is the old output of the failing test and a new one. Note that the base is 1.73 and not 1.74, though it doesn't seem to make much of a difference.
[cargo-show-asm-1-75-ci-failure.zip](https://github.com/pacak/cargo-show-asm/files/13789977/cargo-show-asm-1-75-ci-failure.zip)

**Comment by osiewicz** [comment]
It looks like `main` is now stripped? That seems somewhat plausible as it's defined in `lib.rs`, so who knows..

**Comment by pacak** [comment]
Interestingly enough it works on my machine with 1.75.0

```
% cargo run --release --  --manifest-path sample/Cargo.toml --native --intel
    Finished release [optimized] target(s) in 0.02s
     Running `target/release/cargo-asm --manifest-path sample/Cargo.toml --native --intel`
    Finished release [optimized + debuginfo] target(s) in 0.00s

Try one of those by name or a sequence number
0 "<&T as core::fmt::Display>::fmt" [17]
1 "<sample::MyRngCore as rand_core::block::BlockRngCore>::generate" [18]
2 "core::ptr::drop_in_place<ahash::random_state::DefaultRandomSource>" [7]
3 "hashbrown::raw::RawTable<T,A>::reserve_rehash" [1238]
4 "hashbrown::raw::RawTable<T,A>::reserve_rehash::{{closure}}" [299]
5 "hashbrown::set::HashSet<T,S,A>::insert" [382]
6 "hashbrown::set::HashSet<T>::new" [241]
7 "make_bar" [8]
8 "sample::main" [751]
% rustc --version
rustc 1.75.0 (82e1608df 2023-12-21)
```

Can you try passing `--everything` to check if `sample::main` is there under some other name? Or any ideas how to reproduce it on my end?

**Comment by pacak** [comment]
Stripping shouldn't affect anything - c-s-a doesn't run on compiled code, it asks rustc to dump generated `.s` file.

**Comment by osiewicz** [comment]
Can you try building it with --no-default-features on `sample`? E.g:
`cargo run --release  -- --manifest-path sample/Cargo.toml --native --intel sample::main --rust --no-default-features`
It looks like that makes a difference, as we also test the same scenario *with* default features on sample in CI and that passes even on 1.75.

By stripping I've meant dead code elimination.

**Comment by osiewicz** [comment]
I have a hunch, will try that in the morning though;
With default features turned off, main has a signature of `fn() -> u32` which IIRC isn't a valid function signature for an entry point. With default features turned on, its signature is `fn() -> ()` and that **is** a valid entry point signature.

**Comment by pacak** [comment]
> cargo run --release  -- --manifest-path sample/Cargo.toml --native --intel sample::main --rust --no-default-features

works for me...

CI seem to detect it when it's there if I run with no argument... But fails to find it when it's called.

`fn() -> ()` vs `-> u32` shouldn't really affect anything, in both cases it should be a label like this: `_ZN6sample4main17hecd2c58e95d09ed1E:`.

I'll poke around, will let you know if I manage to reproduce it.

**Comment by osiewicz** [comment]
Well, IMHO the return value should make a difference as to whether the code is stripped altogether or not.

**Comment by pacak** [comment]
Can you attach the raw `.s` file?

**Comment by osiewicz** [comment]
It is in original post in a .zip, together with .s for 1.73.
I am sorry for not providing one for 1.74, but this is what I had at hand.

**Comment by osiewicz** [comment]
Okay, got the jackpot. The return value was indeed not an issue.
The culprit lied in https://github.com/rust-lang/rust/pull/116505
In short, with default features turned off, main was trivial enough to be marked as `inline` function which then made the symbol weak. Since nobody was referencing it, it got stripped away.
Marking main in `default-features=false` config as `#[inline(never)]` or replacing `1 + 1` with a simple `println!("foo")` call makes the test pass again.

**Comment by pacak** [comment]
Good catch!

**Comment by osiewicz** [comment]
Closing as it's fixed with [e7e31ac71fdc8d350226a39ac2be5a4a47c6bac0](https://github.com/pacak/cargo-show-asm/commit/e7e31ac71fdc8d350226a39ac2be5a4a47c6bac0)
